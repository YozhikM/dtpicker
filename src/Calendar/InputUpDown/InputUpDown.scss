.InputUpDown {
  width: 50px;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 3%;

  button {
    border: none;
    background-color: transparent;
    outline: none;
    cursor: pointer;
  }
}
