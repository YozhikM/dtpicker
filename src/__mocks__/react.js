const react = require('react');
// Resolution for requestAnimationFrame not supported in jest error :
// https://github.com/facebook/react/issues/9102#issuecomment-283873039
global.window = global;
window.addEventListener = () => {};
window.requestAnimationFrame = () => {
  throw new Error('requestAnimationFrame is not supported in Node');
};

module.exports = react;

global.spyMount = (component, ...methodNames) => {
  const mount = require('enzyme').mount;
  const wrapper = mount(component);
  const result = [wrapper];
  methodNames.forEach(methodName => {
    result.push(jest.spyOn(wrapper.instance(), methodName));
  });
  wrapper.instance().forceUpdate();
  return result;
};
