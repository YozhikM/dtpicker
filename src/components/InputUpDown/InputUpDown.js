/* @flow */

import React from 'react';
import SvgIcon from '../SvgIcon';
import s from './InputUpDown.scss';

type Props = {|
  value?: number,
  min?: number,
  max?: number,
  circular?: boolean,
  onChange?: (value: number) => void,
  onClickNumber?: (value: number) => void
|};

type State = {|
  value: number,
  isBtnHold: boolean
|};

class InputUpDown extends React.Component<Props, State> {
  state: State;
  holdIntervalId: ?number;

  constructor(props: Props) {
    super(props);
    this.state = {
      value: this.props.value || 0,
      isBtnHold: false
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.value !== this.state.value) {
      this.setState({ value: nextProps.value });
    }
  }

  increment = () => {
    let value = this.state.value + 1;
    if (typeof this.props.max === 'number' && value > this.props.max) {
      value =
        this.props.circular && typeof this.props.min === 'number' ? this.props.min : this.props.max;
    }
    this.setState({ value }, () => {
      if (this.props.onChange) this.props.onChange(value);
    });
  };

  decrement = () => {
    let value = this.state.value - 1;
    if (typeof this.props.min === 'number' && value < this.props.min) {
      value =
        this.props.circular && typeof this.props.max === 'number' ? this.props.max : this.props.min;
    }
    this.setState({ value }, () => {
      if (this.props.onChange) this.props.onChange(value);
    });
  };

  onClickNumber = () => {
    if (this.props.onClickNumber) {
      this.props.onClickNumber(this.state.value);
    }
  };

  onHoldStart = (cb: Function) => {
    this.setState({ isBtnHold: true });
    setTimeout(() => {
      if (this.state.isBtnHold && !this.holdIntervalId) {
        this.holdIntervalId = this.holdIntervalId = setInterval(() => {
          if (this.state.isBtnHold) cb();
        }, 50);
      }
    }, 300);
  };

  onHoldEnd = () => {
    this.setState({ isBtnHold: false });
    if (this.holdIntervalId) {
      clearInterval(this.holdIntervalId);
      this.holdIntervalId = null;
    }
  };

  render() {
    return (
      <div className={s.InputUpDown}>
        <button
          onClick={this.increment}
          onMouseDown={() => this.onHoldStart(this.increment)}
          onTouchStart={() => this.onHoldStart(this.increment)}
          onMouseUp={this.onHoldEnd}
          onTouchEnd={this.onHoldEnd}
        >
          <SvgIcon file="chevron-up" />
        </button>
        <p onClick={this.onClickNumber}>{this.state.value}</p>
        <button
          onClick={this.decrement}
          onMouseDown={() => this.onHoldStart(this.decrement)}
          onTouchStart={() => this.onHoldStart(this.decrement)}
          onMouseUp={this.onHoldEnd}
          onTouchEnd={this.onHoldEnd}
        >
          <SvgIcon file="chevron-down" />
        </button>
      </div>
    );
  }
}

export default InputUpDown;
