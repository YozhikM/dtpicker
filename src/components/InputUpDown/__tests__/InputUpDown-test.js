/* @flow */

import * as React from 'react';
import InputUpDown from '../InputUpDown';

function findUpBtn(wrapper) {
  return wrapper.find('button').at(0);
}

function findDownBtn(wrapper) {
  return wrapper.find('button').at(1);
}

function findP(wrapper) {
  return wrapper.find('p');
}

function findText(wrapper) {
  return wrapper.find('p').text();
}

describe('InputUpDown', () => {
  describe('render()', () => {
    it('basic', () => {
      expect(shallow(<InputUpDown value={59} />)).toMatchSnapshot();
    });
  });

  describe('increment()', () => {
    it('change state', () => {
      const wrapper = shallow(<InputUpDown value={20} />);
      wrapper.instance().increment();
      expect(wrapper.state('value')).toEqual(21);
    });

    it('onChange callback', () => {
      const onChange = jest.fn();
      const wrapper = shallow(<InputUpDown value={20} onChange={onChange} />);
      wrapper.instance().increment();
      expect(onChange).toBeCalledWith(21);
    });

    it('if circular', () => {
      const wrapper = shallow(<InputUpDown value={59} min={10} max={59} circular />);
      wrapper.instance().increment();
      expect(wrapper.state('value')).toEqual(10);
    });

    it('if !circular', () => {
      const wrapper = shallow(<InputUpDown value={59} />);
      wrapper.instance().increment();
      expect(wrapper.state('value')).toEqual(60);
    });

    it('if !circular with max', () => {
      const wrapper = shallow(<InputUpDown value={59} max={59} />);
      wrapper.instance().increment();
      expect(wrapper.state('value')).toEqual(59);
    });
  });

  describe('decrement()', () => {
    it('change state', () => {
      const wrapper = shallow(<InputUpDown value={20} />);
      wrapper.instance().decrement();
      expect(wrapper.state('value')).toEqual(19);
    });

    it('onChange callback', () => {
      const spy = jest.fn();
      const wrapper = shallow(<InputUpDown value={20} onChange={spy} />);
      wrapper.instance().decrement();
      expect(spy).toBeCalledWith(19);
    });

    it('if circular', () => {
      const wrapper = shallow(<InputUpDown value={10} min={10} max={59} circular />);
      wrapper.instance().decrement();
      expect(wrapper.state('value')).toEqual(59);
    });

    it('if !circular', () => {
      const wrapper = shallow(<InputUpDown value={59} />);
      wrapper.instance().decrement();
      expect(wrapper.state('value')).toEqual(58);
    });

    it('if !circular with min', () => {
      const wrapper = shallow(<InputUpDown value={1} min={1} />);
      wrapper.instance().decrement();
      expect(wrapper.state('value')).toEqual(1);
    });
  });

  describe('user interactions', () => {
    it('button up click', () => {
      const [wrapper, spy] = spyMount(<InputUpDown value={20} />, 'increment');
      findUpBtn(wrapper).simulate('click');
      expect(spy).toBeCalled();
      expect(findText(wrapper)).toBe('21');
    });

    it('button up mouseDown', () => {
      const [wrapper, spy] = spyMount(<InputUpDown value={20} />, 'onHoldStart');
      const { increment } = wrapper.instance();
      findUpBtn(wrapper).simulate('mouseDown');
      expect(spy).toBeCalledWith(increment);
      expect(findText(wrapper)).toBe('20');
    });

    it('button up mouseUp', () => {
      const [wrapper, spy] = spyMount(<InputUpDown value={20} />, 'onHoldEnd');
      findUpBtn(wrapper).simulate('mouseUp');
      expect(spy).toBeCalled();
      expect(findText(wrapper)).toBe('20');
    });

    it('button down click', () => {
      const [wrapper, spy] = spyMount(<InputUpDown value={20} />, 'decrement');
      findDownBtn(wrapper).simulate('click');
      expect(spy).toBeCalled();
      expect(findText(wrapper)).toBe('19');
    });

    it('button down mouseDown', () => {
      const [wrapper, spy] = spyMount(<InputUpDown value={20} />, 'onHoldStart');
      const { decrement } = wrapper.instance();
      findDownBtn(wrapper).simulate('mouseDown');
      expect(spy).toBeCalledWith(decrement);
      expect(findText(wrapper)).toBe('20');
    });
  });

  it('componentWillReceiveProps()', () => {
    const wrapper = shallow(<InputUpDown value={1} />);
    wrapper.setProps({ value: 16 });
    expect(wrapper.state('value')).toBe(16);
    expect(findText(wrapper)).toBe('16');
  });

  it('onClickNumber()', () => {
    const spy = jest.fn();
    const wrapper = shallow(<InputUpDown value={20} onClickNumber={spy} />);
    findP(wrapper).simulate('click');
    expect(spy).toBeCalledWith(20);
  });
});
