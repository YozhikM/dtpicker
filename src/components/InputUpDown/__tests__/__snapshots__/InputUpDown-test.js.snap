// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputUpDown render() basic 1`] = `
<div>
  <button
    onClick={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <SvgIcon
      file="chevron-up"
    />
  </button>
  <p
    onClick={[Function]}
  >
    59
  </p>
  <button
    onClick={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <SvgIcon
      file="chevron-down"
    />
  </button>
</div>
`;
